### ISSUE TITLE
Proxy setup issue

### ISSUE THREAD

**Issue by seanng123456789** [wkeeling/selenium-wire#190]
Hi wkeeling,

I am sorry for asking this question, however, I am actually a very newbie for writing program with this kind of technology. The following attach file is my python code, it's a simply example for testing using proxy for crawling. I follow your guides to set a free elite proxy ip and port number. However, the following attached error occur. Sometimes, I can see it opened successfully but still loading, but sometimes just can't load the page.
I wanna know whether I missed setup something or my code is simply incorrect. In addition, I have setup this proxy ip into my chrome settings, and then I open my chrome browser to access this url "https://whatismyipaddress.com/", it can open successfully.(duno if it is any error)
[error.txt](https://github.com/wkeeling/selenium-wire/files/5782960/error.txt)
[sourceCode.txt](https://github.com/wkeeling/selenium-wire/files/5782961/sourceCode.txt)

Thank you so much

**Comment by wkeeling** [comment]
Hi! Thanks for raising this. I think you're suffering from an issue that a number of people have also experienced when connecting using upstream proxy servers that seem to be busy. It seems that Selenium Wire (or rather the lower level code that Selenium Wire depends on) doesn't handle these busy upstream servers terribly well. I'm looking at trying to improve this, but in the meantime the best workaround is to switch to the [mitmproxy backend](https://github.com/wkeeling/selenium-wire#backends) if you are able to. To do that, you'll need to install mitmproxy with:

```
pip install mitmproxy
```
Once installed, set the backend option in Selenium Wire's options to mitmproxy:

```
seleniumwire_options = {
    'backend': 'mitmproxy',
    'proxy': {
        'https': 'https://191.103.219.225:48612'
    },
    'connection_timeout': None  # Never timeout
}
```

**Comment by seanng123456789** [comment]
Thank you for your quick response.
However, when I use mitmproxy, in this statement 
"driver = webdriver.Chrome(options=options, seleniumwire_options=seleniumwire_options, executable_path=r"webDriver\window\chromedriver.exe")"

Error occur:

The page scaping throw exception. Url: https://www.facebook.com/familyfeasthk
[WinError 2] The system cannot find the file specified
[WinError 2] The system cannot find the file specified
Traceback (most recent call last):
  File "C:\sean\workspace_running\WebCrawler\WebMonitor\src\Test.py", line 59, in <module>
    driver.quit()
NameError: name 'driver' is not defined

**Comment by wkeeling** [comment]
Are you able to re-post the code (with mitmproxy) and the full traceback?

**Comment by wkeeling** [comment]
@seanng123456789 guessing here, but the issue is possibly happening because `webDriver\window\chromedriver.exe` isn't found, so the webdriver is never instantiated, which means the final `driver.quit()` call then falls over with a `NameError`.

I would do a couple of things: first use the absolute path in `executable_path` - e.g. `C:\full\path\to\webDriver\window\chromedriver.exe` and secondly, check that the `driver` is not `None` before calling `quit()`, so putting it all together:
```
from seleniumwire import webdriver
from random import randrange
import time

profiles_urls = "https://whatismyipaddress.com/"

for profile_url in profiles_urls:

	
	headers = {
		"Upgrade-Insecure-Requests": "1",
		"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
		"Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
		"Sec-Fetch-Site": "none",
		"Sec-Fetch-Mode": "navigate",
		"Sec-Fetch-Dest": "document",
		"Accept-Encoding": r"gzip, deflate, br",
		"Accept-Language": "en-US;q=0.9"
			  }
	
        exe_path = r'C:\full\path\to\webDriver\window\chromedriver.exe'
        driver = None
	try:
		#for setting proxies into selenium wire options
		seleniumwire_options = {
                    'backend': 'mitmproxy',
		    'proxy': {
		        'https': 'https://191.103.219.225:48612'
		    },
		    'connection_timeout': None  # Never timeout
		}
		
		#for setting headless into selenium options
		options = webdriver.ChromeOptions()
		options.headless = False
# 		options.add_argument('--lang=en')

		driver = webdriver.Chrome(options=options, seleniumwire_options=seleniumwire_options, executable_path=exe_path)
# 		driver = webdriver.Chrome(options=options, executable_path=exe_path)
		driver.header_overrides = headers
		driver.get(profile_url)
		time.sleep(randrange(6, 15))
# 		request = driver.wait_for_request(profile_url, timeout=randrange(20, 30))
# 		print(request.headers)
		print(driver.page_source)
				
	except Exception as e:
		print("\nThe page scaping throw exception. Url: " + profile_url + "\n" + str(e))
		
		print(str(e))
	
        if driver is not None:
	    driver.quit()
```

**Comment by seanng123456789** [comment]
All error:
The page scaping throw exception. Url: https://www.facebook.com/familyfeasthk
[WinError 2] The system cannot find the file specified
[WinError 2] The system cannot find the file specified
Traceback (most recent call last):
  File "C:\sean\workspace_running\WebCrawler\WebMonitor\src\Test.py", line 59, in <module>
    driver.quit()
NameError: name 'driver' is not defined

[source code.txt](https://github.com/wkeeling/selenium-wire/files/5787389/source.code.txt)

**Comment by seanng123456789** [comment]
let me try again then, thank you

**Comment by seanng123456789** [comment]
I try it, it seems not related to the path. Once I comment out the mitmproxy backend, it can execute, but if I use the mitmproxy backend, the above error will occur

**Comment by wkeeling** [comment]
OK interesting. It would be good to see the full traceback as currently the exception handler is just printing out the message with `print(str(e))`. Can you add `raise e` afterwards:

```
except Exception as e:
    print("\nThe page scaping throw exception. Url: " + profile_url + "\n" + str(e))
		
    print(str(e))

    raise e  # Add this
```
The exception should then be shown in the console.

**Comment by seanng123456789** [comment]
It's really sorry for my skill set. 

Here is the full stack trace:

Traceback (most recent call last):
  File "C:\sean\workspace_running\WebCrawler\WebMonitor\src\Test.py", line 58, in <module>

The page scaping throw exception. Url: https://www.facebook.com/familyfeasthk
[WinError 2] The system cannot find the file specified
[WinError 2] The system cannot find the file specified
    raise e
  File "C:\sean\workspace_running\WebCrawler\WebMonitor\src\Test.py", line 45, in <module>
    driver = webdriver.Chrome(options=options, seleniumwire_options=seleniumwire_options, executable_path=r"C:\sean\workspace_running\WebCrawler\WebMonitor\src\webDriver\window\chromedriver.exe")
  File "C:\sean\venv\lib\site-packages\seleniumwire\webdriver\browser.py", line 65, in __init__
    addr, port = self._client.create_proxy(
  File "C:\sean\venv\lib\site-packages\seleniumwire\proxy\client.py", line 69, in create_proxy
    self._proxy = mitmproxy.start(addr, port, options)
  File "C:\sean\venv\lib\site-packages\seleniumwire\proxy\mitmproxy.py", line 54, in start
    proxy = subprocess.Popen([
  File "C:\Program Files\Python38\lib\subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Program Files\Python38\lib\subprocess.py", line 1307, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified

**Comment by wkeeling** [comment]
Thanks @seanng123456789 - it looks like you're running an older version of Selenium Wire which starts the backend in a separate subprocess, and for some reason it can't find the backend executable. 

From v3.0.0 onwards however, Selenium Wire runs the backend in the same process and so you shouldn't hit this issue. Are you able to upgrade and then re-try? To upgrade:

```
pip install selenium-wire --upgrade
```

**Comment by seanng123456789** [comment]
ok thanks a lot

**Comment by seanng123456789** [comment]
It seems work for the basic settings, however, I dig some for information for the following error, it seems related to mitmproxy certificate which is rejected by the website, can I do something?

127.0.0.1:50945: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50947: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50943: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50946: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50944: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50961: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50962: Client Handshake failed. The client may not trust the proxy's certificate for static.xx.fbcdn.net.
127.0.0.1:50937: Client Handshake failed. The client may not trust the proxy's certificate for www.facebook.com.

**Comment by wkeeling** [comment]
Glad to hear it's working better @seanng123456789  

The best way to resolve the certificate warning is to install the certificate into Chrome. You should be able to find the certificate in the `.mitmproxy` folder in your home folder. For example:

```
C:\Users\sean\.mitmproxy\mitmproxy-ca-cert.p12
```
You need the `.p12` certificate since you're running on Windows. To install the certificate, you can follow these instructions:

https://web.archive.org/web/20160612045445/http://windows.microsoft.com/en-ca/windows/import-export-certificates-private-keys#1TC=windows-7

**Comment by seanng123456789** [comment]
thank you so much let me try

**Comment by seanng123456789** [comment]
![Capture](https://user-images.githubusercontent.com/74837842/104105269-f538de80-52e7-11eb-95a5-957db0af46ad.PNG)
The instruction didn't mention which kind of folder I should choose, so when the installation I choose automatically choose the folder defined by the type of the certificate. I don't it is correct or not.
After these steps, the above error still occur.

**Comment by wkeeling** [comment]
I believe the certificate should be installed in "Trusted Root Certification Authorities". Let me know if you still have issues.

**Comment by seanng123456789** [comment]
hi thanks for your kindly help, I think I won't need to use mitmproxy btw now. But for this issue, after imported the cert in trusted root certification authorities, the following error still occur.
[error.txt](https://github.com/wkeeling/selenium-wire/files/5824891/error.txt)
[sourceCode.txt](https://github.com/wkeeling/selenium-wire/files/5824892/sourceCode.txt)

Anyway, thanks for your help. For your information.
